ArduCopter/events.cpp

```diff
@@ -4,152 +4,235 @@
+        action = FS_THR_DISABLED;
+
+    } else if (flightmode->is_landing() && ((battery.has_failsafed() && battery.get_highest_failsafe_priority() <= FAILSAFE_LAND_PRIORITY))) {
+        // Allow landing to continue when battery failsafe requires it (not a user option)
+        gcs().send_text(MAV_SEVERITY_WARNING, "RADIO + BATTERY FAILSAFE - CONTINUING LANDING");
+        action = FS_THR_ENABLED_ALWAYS_LAND;
+
+    } else if (flightmode->is_landing() && failsafe_option(FailsafeOption::THR_LAND_CONTINUE)) {
+        // Allow landing to continue when FS_OPTIONS is set to continue landing
+        gcs().send_text(MAV_SEVERITY_WARNING, "RADIO FAILSAFE - CONTINUING LANDING");
+        action = FS_THR_ENABLED_ALWAYS_LAND;
+
+    } else if (control_mode == Mode::Number::AUTO && failsafe_option(FailsafeOption::THR_MISSION_CONTINUE)) {
+        // Allow mission to continue when FS_OPTIONS is set to continue mission
+        gcs().send_text(MAV_SEVERITY_WARNING, "RADIO FAILSAFE - CONTINUING AUTO MODE MISSION");       
+        action = FS_THR_DISABLED;
+
+    } else if ((control_mode == Mode::Number::GUIDED || control_mode == Mode::Number::GUIDED_NOGPS) &&
+      failsafe_option(FailsafeOption::THR_GUIDED_CONTINUE) && (g.failsafe_gcs != FS_GCS_DISABLED)) {
+        // Allow guided mode to continue when FS_THR_OPTION is set to continue in guided mode.  Only if the GCS failsafe is enabled.
+        gcs().send_text(MAV_SEVERITY_WARNING, "RADIO FAILSAFE - CONTINUING GUIDED MODE");
+        action = FS_THR_DISABLED;
     } else {
-        if (control_mode == Mode::Number::AUTO && g.failsafe_throttle == FS_THR_ENABLED_CONTINUE_MISSION) {
-            // continue mission
-        } else if (flightmode->is_landing() &&
-                   battery.has_failsafed() &&
-                   battery.get_highest_failsafe_priority() <= FAILSAFE_LAND_PRIORITY) {
-            // continue landing or other high priority failsafes
-        } else {
-            if (g.failsafe_throttle == FS_THR_ENABLED_ALWAYS_RTL) {
-                set_mode_RTL_or_land_with_pause(MODE_REASON_RADIO_FAILSAFE);
-            } else if (g.failsafe_throttle == FS_THR_ENABLED_CONTINUE_MISSION) {
-                set_mode_RTL_or_land_with_pause(MODE_REASON_RADIO_FAILSAFE);
-            } else if (g.failsafe_throttle == FS_THR_ENABLED_ALWAYS_SMARTRTL_OR_RTL) {
-                set_mode_SmartRTL_or_RTL(MODE_REASON_RADIO_FAILSAFE);
-            } else if (g.failsafe_throttle == FS_THR_ENABLED_ALWAYS_SMARTRTL_OR_LAND) {
-                set_mode_SmartRTL_or_land_with_pause(MODE_REASON_RADIO_FAILSAFE);
-            } else { // g.failsafe_throttle == FS_THR_ENABLED_ALWAYS_LAND
-                set_mode_land_with_pause(MODE_REASON_RADIO_FAILSAFE);
-            }
-        }
+        gcs().send_text(MAV_SEVERITY_WARNING, "RADIO FAILSAFE");
     }
 
-    AP::logger().Write_Error(LogErrorSubsystem::FAILSAFE_RADIO, LogErrorCode::FAILSAFE_OCCURRED);
-
+    // Execute the specified action
+    switch(action) {
```

Comment:
nitpick: some missing spaces here and there like between "switch" and the following "("

Reply:
Fixed